content/blog/2025-04-02-introducing-bloom-filters.md

```diff
@@ -0,0 +1,188 @@
+
+1. Perform a memory check prior to bloom filter creation
+
+We can use the `VALIDATESCALETO` option of the `BF.INSERT` command to perform a validation whether the filter is within the memory limit. If it is not within the limits, the command is rejected. In the example below, we see that filter1 cannot scale out and reach the capacity of 26214301 due to the memory limit. However, it can scale out and reach a capacity of 26214300.
+```
+127.0.0.1:6379> BF.INSERT filter1 VALIDATESCALETO 26214301
+(error) ERR provided VALIDATESCALETO causes bloom object to exceed memory limit
+127.0.0.1:6379> BF.INSERT filter1 VALIDATESCALETO 26214300 ITEMS item1
+1) (integer) 1
+```
+2. Check the maximum capacity that an existing scalable bloom filter can expand to
+
+We can use the `BF.INFO` command to find out the maximum capacity that the scalable bloom filter can expand to hold. In this case, we can see the filter can hold 26214300 items (after scaling out until the memory limit).
+```
+127.0.0.1:6379> BF.INFO filter1 MAXSCALEDCAPACITY
+(integer) 26214300
+```
+
+To get an idea of what the memory usage looks like for the max capacity of an individual non scaling filter, we have a table below. With a 128MB limit and default false positive rate, we can create a bloom filter with 112M as the capacity. With a 512MB limit, a bloom filter can hold 448M items.
+
+<table width="100%" border="1" style="border-collapse: collapse; border: 1px solid black" cellpadding="8">
+<tr>
+<th>Non Scaling Filter - Capacity</th>
+<th>FP Rate</th>
+<th>Memory Usage (MB)</th>
+<th>Notes</th>
+</tr>
+<tr>
+<td>112M</td>
+<td>0.01</td>
+<td>~128</td>
+<td>Default FP Rate and Default Memory Limit</td>
+</tr>
+<tr>
+<td>74M</td>
+<td>0.001</td>
+<td>~128</td>
+<td>Custom FP Rate and Default Memory Limit</td>
+</tr>
+<tr>
+<td>448M</td>
+<td>0.01</td>
+<td>~512</td>
+<td>Default FP Rate and Custom Memory Limit</td>
+</tr>
+<tr>
+<td>298M</td>
+<td>0.001</td>
+<td>~512</td>
+<td>Custom FP Rate and Custom Memory Limit</td>
+</tr>
+</table>
+
+## Performance 
+
+The bloom commands which involve adding items or checking the existence of items have a time complexity of O(N * K) where N is the number of hash functions used by the bloom filter and K is the number of elements being inserted. This means that both `BF.ADD` and `BF.EXISTS` are both O(N) as they only operate on one item.
+
+In case of scalable bloom filters, with every scale out, we increase the number of checks (using hash functions of each sub filter) performed during any add / exists operation. For this reason, it is recommended that users choose a capacity and expansion rate after evaluating the use case / workload to avoid several scale outs and reduce the number of checks.
```

Comment:
How does a user choose the right values? This isn't very prescriptive for the average user.

Reply:
I made changes here: https://github.com/valkey-io/valkey-io.github.io/pull/229/commits/29a1e1b1d7e022b9cf2b4d8f29fb9f35d7058b92